### PR TITLE
Add NewHope.

### PR DESCRIPTION
**Pull request by aparent**
First pass at adding new-hope.  I modified reference version to fit the OQS API (including replacing random number generation with OQS_RAND).

### CODE REVIEW

File: src/kex_rlwe_newhope/error_correction.c [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	t += g(xi2);
+	t += g(xi3);
+
+	t -= 8 * PARAM_Q;
+	t >>= 31;
+	return t & 1;
+}
+
+
+void helprec(poly *c, const poly *v, OQS_RAND *oqs_rand) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/error_correction.c]: Global symbol

File: src/kex_rlwe_newhope/error_correction.c [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+		c->coeffs[  0 + i] = (v_tmp[0] -   v_tmp[3]) & 3;
+		c->coeffs[256 + i] = (v_tmp[1] -   v_tmp[3]) & 3;
+		c->coeffs[512 + i] = (v_tmp[2] -   v_tmp[3]) & 3;
+		c->coeffs[768 + i] = (   -k    + 2 * v_tmp[3]) & 3;
+	}
+}
+
+
+void rec(unsigned char *key, const poly *v, const poly *c) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/error_correction.c]: Global symbol

File: src/kex_rlwe_newhope/fips202.c [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+	(uint64_t)0x8000000000000080ULL,
+	(uint64_t)0x000000000000800aULL,
+	(uint64_t)0x800000008000000aULL,
+	(uint64_t)0x8000000080008081ULL,
+	(uint64_t)0x8000000000008080ULL,
+	(uint64_t)0x0000000080000001ULL,
+	(uint64_t)0x8000000080008008ULL
+};
+
+void KeccakF1600_StatePermute(uint64_t *state) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/fips202.c]: Global symbol

File: src/kex_rlwe_newhope/fips202.c [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+		for (i = 0; i < (r >> 3); i++) {
+			store64(h + 8 * i, s[i]);
+		}
+		h += r;
+		nblocks--;
+	}
+}
+
+
+void shake128_absorb(uint64_t *s, const unsigned char *input, unsigned int inputByteLen) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/fips202.c]: Global symbol

File: src/kex_rlwe_newhope/fips202.c [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+	keccak_absorb(s, SHAKE128_RATE, input, inputByteLen, 0x1F);
+}
+
+
+void shake128_squeezeblocks(unsigned char *output, unsigned long long nblocks, uint64_t *s) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/fips202.c]: Global symbol

File: src/kex_rlwe_newhope/fips202.c [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+	keccak_squeezeblocks(output, nblocks, s, SHAKE128_RATE);
+}
+
+
+void shake128(unsigned char *output, unsigned int outputByteLen, const unsigned char *input, unsigned int inputByteLen) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/fips202.c]: Global symbol

File: src/kex_rlwe_newhope/fips202.c [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+	uint64_t s[25];
+	assert(!(outputByteLen % SHAKE128_RATE));
+	shake128_absorb(s, input, inputByteLen);
+	shake128_squeezeblocks(output, outputByteLen / SHAKE128_RATE, s);
+}
+
+
+void sha3256(unsigned char *output, const unsigned char *input, unsigned int inputByteLen) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/fips202.c]: Global symbol

File: src/kex_rlwe_newhope/ntt.c [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	19, 531, 275, 787, 147, 659, 403, 915, 83, 595, 339, 851, 211, 723, 467, 979, 51, 563, 307, 819, 179, 691, 435, 947, 115, 627, 371, 883, 243, 755, 499, 1011,
+	11, 523, 267, 779, 139, 651, 395, 907, 75, 587, 331, 843, 203, 715, 459, 971, 43, 555, 299, 811, 171, 683, 427, 939, 107, 619, 363, 875, 235, 747, 491, 1003,
+	27, 539, 283, 795, 155, 667, 411, 923, 91, 603, 347, 859, 219, 731, 475, 987, 59, 571, 315, 827, 187, 699, 443, 955, 123, 635, 379, 891, 251, 763, 507, 1019,
+	7, 519, 263, 775, 135, 647, 391, 903, 71, 583, 327, 839, 199, 711, 455, 967, 39, 551, 295, 807, 167, 679, 423, 935, 103, 615, 359, 871, 231, 743, 487, 999,
+	23, 535, 279, 791, 151, 663, 407, 919, 87, 599, 343, 855, 215, 727, 471, 983, 55, 567, 311, 823, 183, 695, 439, 951, 119, 631, 375, 887, 247, 759, 503, 1015,
+	15, 527, 271, 783, 143, 655, 399, 911, 79, 591, 335, 847, 207, 719, 463, 975, 47, 559, 303, 815, 175, 687, 431, 943, 111, 623, 367, 879, 239, 751, 495, 1007,
+	31, 543, 287, 799, 159, 671, 415, 927, 95, 607, 351, 863, 223, 735, 479, 991, 63, 575, 319, 831, 191, 703, 447, 959, 127, 639, 383, 895, 255, 767, 511, 1023
+};
+
+void bitrev_vector(uint16_t *poly) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/ntt.c]: Global symbol

File: src/kex_rlwe_newhope/ntt.c [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+		r = bitrev_table[i];
+		if (i < r) {
+			tmp = poly[i];
+			poly[i] = poly[r];
+			poly[r] = tmp;
+		}
+	}
+}
+
+void mul_coefficients(uint16_t *poly, const uint16_t *factors) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/ntt.c]: Global symbol

File: src/kex_rlwe_newhope/ntt.c [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	unsigned int i;
+
+	for (i = 0; i < PARAM_N; i++)
+		poly[i] = montgomery_reduce((poly[i] * factors[i]));
+}
+
+
+
+/* GS_bo_to_no; omegas need to be in Montgomery domain */
+void ntt(uint16_t *a, const uint16_t *omega) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/ntt.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 11]
```diff
@@ -0,0 +1,129 @@
+#include "poly.h"
+#include "ntt.h"
+#include "reduce.h"
+#include "fips202.h"
+
+void poly_frombytes(poly *r, const unsigned char *a) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+	int i;
+	for (i = 0; i < PARAM_N / 4; i++) {
+		r->coeffs[4 * i + 0] =                               a[7 * i + 0]        | (((uint16_t)a[7 * i + 1] & 0x3f) << 8);
+		r->coeffs[4 * i + 1] = (a[7 * i + 1] >> 6) | (((uint16_t)a[7 * i + 2]) << 2) | (((uint16_t)a[7 * i + 3] & 0x0f) << 10);
+		r->coeffs[4 * i + 2] = (a[7 * i + 3] >> 4) | (((uint16_t)a[7 * i + 4]) << 4) | (((uint16_t)a[7 * i + 5] & 0x03) << 12);
+		r->coeffs[4 * i + 3] = (a[7 * i + 5] >> 2) | (((uint16_t)a[7 * i + 6]) << 6);
+	}
+}
+
+void poly_tobytes(unsigned char *r, const poly *p) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+		r[7 * i + 1] = (t0 >> 8) | (t1 << 6);
+		r[7 * i + 2] = (t1 >> 2);
+		r[7 * i + 3] = (t1 >> 10) | (t2 << 4);
+		r[7 * i + 4] = (t2 >> 4);
+		r[7 * i + 5] = (t2 >> 12) | (t3 << 2);
+		r[7 * i + 6] = (t3 >> 6);
+	}
+}
+
+void poly_uniform(poly *a, const unsigned char *seed) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+		if (pos > SHAKE128_RATE * nblocks - 2) {
+			nblocks = 1;
+			shake128_squeezeblocks((unsigned char *) buf, nblocks, state);
+			pos = 0;
+		}
+	}
+}
+
+
+void poly_getnoise(poly *r, OQS_RAND *rand) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+		d = 0;
+		for (j = 0; j < 8; j++)
+			d += (t >> j) & 0x01010101;
+		a = ((d >> 8) & 0xff) + (d & 0xff);
+		b = (d >> 24) + ((d >> 16) & 0xff);
+		r->coeffs[i] = a + PARAM_Q - b;
+	}
+}
+
+void poly_pointwise(poly *r, const poly *a, const poly *b) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+	int i;
+	uint16_t t;
+	for (i = 0; i < PARAM_N; i++) {
+		t       = montgomery_reduce(3186 * b->coeffs[i]); /* t is now in Montgomery domain */
+		r->coeffs[i] = montgomery_reduce(a->coeffs[i] * t); /* r->coeffs[i] is back in normal domain */
+	}
+}
+
+void poly_add(poly *r, const poly *a, const poly *b) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+	int i;
+	for (i = 0; i < PARAM_N; i++)
+		r->coeffs[i] = barrett_reduce(a->coeffs[i] + b->coeffs[i]);
+}
+
+void poly_ntt(poly *r) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/poly.c [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+	mul_coefficients(r->coeffs, psis_bitrev_montgomery);
+	ntt((uint16_t *)r->coeffs, omegas_montgomery);
+}
+
+void poly_invntt(poly *r) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/poly.c]: Global symbol

File: src/kex_rlwe_newhope/precomp.c [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
...2975, 11726, 2744, 9283, 10092, 5067, 12171, 2476, 3748, 11336, 6522, 827, 9452, 5374, 12159, 7935, 3296, 3949, 9893, 4452, 10908, 2525, 3584, 8112, 8011, 10616, 4989, 6958, 11809, 9447, 12280, 1022, 11950, 9821, 11745, 5791, 5092, 2089, 9005, 2881, 3289, 2013, 9048, 729, 7901, 1260, 5755, 4632, 11955, 2426, 10593, 1428, 4890, 5911, 3932, 9558, 8830, 3637, 5542, 145, 5179, 8595, 3707, 10530, 355, 3382, 4231, 9741, 1207, 9041, 7012, 1168, 10146, 11224, 4645, 11885, 10911, 10377, 435, 7952, 4096, 493, 9908, 6845, 6039, 2422, 2187, 9723, 8643, 9852, 9302, 6022, 7278, 1002, 4284, 5088, 1607, 7313, 875, 8509, 9430, 1045, 2481, 5012, 7428, 354, 6591, 9377, 11847, 2401, 1067, 7188, 11516, 390, 8511, 8456, 7270, 545, 8585, 9611, 12047, 1537, 4143, 4714, 4885, 1017, 5084, 1632, 3066, 27, 1440, 8526, 9273, 12046, 11618, 9289, 3400, 9890, 3136, 7098, 8758, 11813, 7384, 3985, 11869, 6730, 10745, 10111, 2249, 4048, 2884, 11136, 2126, 1630, 9103, 5407, 2686, 9042, 2969, 8311, 9424, 9919, 8779, 5332, 10626, 1777, 4654, 10863, 7351, 3636, 9585, 5291, 8374, 2166, 4919, 12176, 9140, 12129, 7852, 12286, 4895, 10805, 2780, 5195, 2305, 7247, 9644, 4053, 10600, 3364, 3271, 4057, 4414, 9442, 7917, 2174};
```
Comment (dstebila) [inline src/kex_rlwe_newhope/precomp.c]: Global symbol

File: src/kex_rlwe_newhope/precomp.c [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
...8889, 3000, 671, 243, 3016, 3763, 10849, 12262, 9223, 10657, 7205, 11272, 7404, 7575, 8146, 10752, 242, 2678, 3704, 11744, 5019, 3833, 3778, 11899, 773, 5101, 11222, 9888, 442, 2912, 5698, 11935, 4861, 7277, 9808, 11244, 2859, 3780, 11414, 4976, 10682, 7201, 8005, 11287, 5011, 6267, 2987, 2437, 3646, 2566, 10102, 9867, 6250, 5444, 2381, 11796, 8193, 4337, 11854, 1912, 1378, 404, 7644, 1065, 2143, 11121, 5277, 3248, 11082, 2548, 8058, 8907, 11934, 1759, 8582, 3694, 7110, 12144, 6747, 8652, 3459, 2731, 8357, 6378, 7399, 10861, 1696, 9863, 334, 7657, 6534, 11029, 4388, 11560, 3241, 10276, 9000, 9408, 3284, 10200, 7197, 6498, 544, 2468, 339, 11267, 9, 2842, 480, 5331, 7300, 1673, 4278, 4177, 8705, 9764, 1381, 7837, 2396, 8340, 8993, 4354, 130, 6915, 2837, 11462, 5767, 953, 8541, 9813, 118, 7222, 2197, 3006, 9545, 563, 9314, 2625, 11340, 4821, 2639, 7266, 5828, 6561, 7698, 3328, 6512, 1351, 7311, 6553, 8155, 1305, 722, 5146, 4043, 12288, 10810, 2545, 3621, 8747, 8785, 1646, 1212, 5860, 3195, 7203, 10963, 3201, 3014, 955, 11499, 9970, 11119, 3135, 3712, 7443, 9542, 7484, 8736, 9995, 11227, 1635, 9521, 1177, 8034, 140, 10436, 11563, 7678, 4320, 11289, 9198, 12208, 2963, 7393, 2366, 9238};
```
Comment (dstebila) [inline src/kex_rlwe_newhope/precomp.c]: Global symbol

File: src/kex_rlwe_newhope/precomp.c [hunk 21]
```diff
@@ -0,0 +1,11 @@
...
...1755, 7273, 1039, 1904, 272, 3550, 9285, 3082, 5707, 6082, 4380, 7648, 11626, 5172, 4250, 9385, 8363, 8217, 4685, 5936, 848, 8899, 6538, 934, 1889, 3781, 9318, 10109, 10222, 6727, 961, 5404, 772, 5377, 9546, 8386, 1198, 8949, 3034, 2189, 7335, 4559, 5918, 2601, 10905, 5069, 9502, 3113, 7467, 8089, 11689, 5181, 9518, 8382, 2953, 3933, 4073, 4093, 7607, 8109, 2914, 5683, 4323, 11151, 1593, 10761, 6804, 972, 3650, 2277, 5592, 4310, 7638, 9869, 4921, 703, 1856, 9043, 4803, 9464, 1352, 8971, 11815, 5199, 7765, 6376, 4422, 7654, 2849, 407, 8836, 6529, 7955, 2892, 9191, 1313, 10721, 12065, 12257, 1751, 9028, 8312, 2943, 2176, 3822, 546, 78, 8789, 11789, 10462, 12028, 6985, 4509, 9422, 1346, 5459, 4291, 613, 10621, 6784, 9747, 3148, 7472, 2823, 5670, 810, 7138, 8042, 4660, 7688, 6365, 6176, 6149, 2634, 5643, 9584, 10147, 11983, 5223, 9524, 11894, 10477, 8519, 1217, 3685, 2282, 326, 10580, 3267, 7489, 4581, 2410, 5611, 11335, 6886, 8006, 8166, 11700, 3427, 11023, 8597, 10006, 3185, 455, 65, 5276, 7776, 4622, 5927, 7869, 9902, 11948, 5218, 2501, 5624, 2559, 10899, 1557, 1978, 10816, 10323, 8497, 4725, 675, 1852, 10798, 12076, 10503, 3256, 9243, 3076, 2195, 10847, 12083, 10504, 12034, 10497};
```
Comment (dstebila) [inline src/kex_rlwe_newhope/precomp.c]: Global symbol

File: src/kex_rlwe_newhope/reduce.c [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+/* Incomplete-reduction routines; for details on allowed input ranges
+ * and produced output ranges, see the description in the paper:
+ * https://cryptojedi.org/papers/#newhope */
+
+
+static const uint32_t qinv = 12287; // -inverse_mod(p,2^18)
+static const uint32_t rlog = 18;
+
+uint16_t montgomery_reduce(uint32_t a) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/reduce.c]: Global symbol

File: src/kex_rlwe_newhope/reduce.c [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+	u = (a * qinv);
+	u &= ((1 << rlog) - 1);
+	u *= PARAM_Q;
+	a = a + u;
+	return a >> 18;
+}
+
+
+uint16_t barrett_reduce(uint16_t a) {
```
Comment (dstebila) [inline src/kex_rlwe_newhope/reduce.c]: Global symbol

**Comment by aparent** [comment]
Sorry I pushed to the wrong repo.  I intended for that to be checked first. That check should fix up the global namespace though.

**Comment by dstebila** [comment]
No worries.  I'll review it over the weekend.

**Comment by dstebila** [comment]
Note to self: things to check / add:

- prettyprint
- global namespace
- license
- README
- project website
- OpenSSL integration

**Pull request merged**